### ISSUE TITLE
Improve function `Math.tryAdd`

### ISSUE THREAD

**Issue by barakman** [OpenZeppelin/openzeppelin-contracts#4857]
**🧐 Motivation**
- Reduce the gas consumption of function `Math.tryAdd`
- Consolidate the output of function `Math.tryAdd`
- Expand the usability of function `Math.tryAdd`
 
**📝 Details**
In addition to the sum of two input values, this function returns an indication of whether or not the calculation is safe:
- True if an overflow does not occur during the calculation
- False if an overflow does occur during the calculation

However, in the case of an overflow, it also returns the calculated sum as zero instead of as the actual (overflowing) result.

The implications of this are:
1. Output redundancy, as both returned values provide the exact same information, namely - an overflow has occurred
2. Output deficiency, as the 2nd returned value does not provide the overflowing result, which may nevertheless be useful
3. Increased gas consumption

My point in bullet 2 above is that the information has already been calculated, so why not return it to the user and let them decide what to do with it? Perhaps you've decided to return 0 as some sort of "safety net" in case of erroneous usage by the caller of the function, but under such scenario, I cannot quite see how 0 would be "safer" than any other value.

This is [the current implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/math/Math.sol#L25-L31):
```
function tryAdd(uint256 a, uint256 b) internal pure returns (bool success, uint256 result) {
    unchecked {
        uint256 c = a + b;
        if (c < a) return (false, 0);
        return (true, c);
    }
}
```
And this is my suggestion for an alternative implementation:
```
function tryAdd(uint256 a, uint256 b) internal pure returns (bool success, uint256 result) {
    unchecked {
        result = a + b;
        success = result >= a;
     }
}
```
Gas measurement and comparison conducted as follows:
- Compiler version 0.8.20
- Optimizer enabled with 200 runs

Contract code:
```
// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity 0.8.20;

abstract contract Contract {
    uint256 public gasUsed;

    function test(uint256 a, uint256 b) external returns (bool success, uint256 result) {
        uint256 gasLeft = gasleft();
        (success, result) = tryAdd(a, b);
        gasUsed = gasLeft - gasleft();
    }

    function tryAdd(uint256, uint256) internal virtual pure returns (bool, uint256);
}

contract Contract1 is Contract {
    function tryAdd(uint256 a, uint256 b) internal override pure returns (bool, uint256) {
        unchecked {
            uint256 c = a + b;
            if (c < a) return (false, 0);
            return (true, c);
        }
    }
}

contract Contract2 is Contract {
    function tryAdd(uint256 a, uint256 b) internal override pure returns (bool success, uint256 result) {
        unchecked {
            result = a + b;
            success = result >= a;
        }
    }
}
```
Truffle test code:
```
const Contract1 = artifacts.require("Contract1");
const Contract2 = artifacts.require("Contract2");

const MAX = web3.utils.toBN(1).shln(256);

describe.only("test", () => {
    it("test", async () => {
        console.log();
        for (let i = 0; i < 5; i++) {
            for (let j = 0; j < 5; j++) {
                await test(MAX.addn(i).mod(MAX), MAX.addn(j).mod(MAX));
                await test(MAX.addn(i).mod(MAX), MAX.subn(j).mod(MAX));
                await test(MAX.subn(i).mod(MAX), MAX.addn(j).mod(MAX));
                await test(MAX.subn(i).mod(MAX), MAX.subn(j).mod(MAX));
            }
        }
    });

    async function test(a, b) {
        const gasUsed = [];
        for (const Contract of [Contract1, Contract2]) {
            const contract = await Contract.new();
            const receipt = await contract.test(a, b);
            gasUsed.push(await contract.gasUsed());
        }
        console.log(`${gasUsed[0]}, ${gasUsed[1]} (a = ${a}, b = ${b})`);
        assert(gasUsed[0].gte(gasUsed[1]));
    }
});
```
I've also tested with `[Contract2, Contract1]`, in order to make sure that the order of execution doesn't impact the results.

TLDR gas measurement:
- Your current implementation: either 117 or 131 gas per function call
- My suggested implementation: 38 gas per function call

It is slightly surprising, since my implementation replaces your implementation's usage of `<` with `>=`.
Last time I checked, the latter was slightly more expensive than the former (5 gas vs 3 gas).
Perhaps it is your forcing of `0` which makes your implementation ultimately more expensive than my implementation.
Or perhaps it is because you are not using the function-declaration's return variables `(bool success, uint256 result)`.

One important note to point out is that if you choose to accept this suggestion, then it will break any code which happens to rely on both returned value for detecting an overflow, i.e., `if (success == false && result == 0)`.

And despite the fact that such logic is generally wrong, it should probably be emphasized in the related release notes.

A similar approach can probably be used also for function `trySub` and for function `tryMul`.

Thank you :)

**Comment by Amxx** [comment]
Hello @barakman and thank you for this interresting proposal

> One important note to point out is that if you choose to accept this suggestion, then it will break any code which happens to rely on both returned value for detecting an overflow, i.e., `if (success == false && result == 0)`.

I think this is an issue. We care a lot about backward compatibility, and even though we don't know if anyone relies on the result being 0, some users may rely on that.

**Comment by Amxx** [comment]
If there is a way to reduce the gas cost while keeping the current behavior (result == 0 in case of overflow), we would obviously be open to that.

**Comment by barakman** [comment]
> Hello @barakman and thank you for this interresting proposal
> 
> > One important note to point out is that if you choose to accept this suggestion, then it will break any code which happens to rely on both returned value for detecting an overflow, i.e., `if (success == false && result == 0)`.
> 
> I think this is an issue. We care a lot about backward compatibility, and even though we don't know if anyone relies on the result being 0, some users may rely on that.
> 
> This check could make sens if the some project does
> 
> ```solidity
> (uint256 result, bool success) = tryAdd(x, y);
> if (result > 0) {
>     // do stuff
> } else require(success, "overflow");
> ```

As I explained, relying on "False AND Zero" as an indication of overflow, is generally wrong.
Even more wrong, is to rely only on "Zero" as an indication of overflow, and that's **regardless** of my suggestion.
This is because 0 + 0 does not overflow.

In other words, the function may return `(true, 0)` and it may return `(false, 0)`.
The latter indicates an overflow, while the former indicates no overflow.

You code snippet above is actually a good example of how wrong it is to rely on 0 alone as an indication of overflow.

Thank you for your quick response :)

**Comment by ernestognw** [comment]
Thanks for these great observations @barakman!

> In other words, the function may return (true, 0) and it may return (false, 0).
> The latter indicates an overflow, while the former indicates no overflow.

Right this is a great point and I also think it may not be a great example of how to not use the library.

I think the issue is that the meaning of the return value is just different semantically. Regardless of whether or not the correct usage is to rely on `0`. I'd expect the most popular use case of `tryAdd` would be to use both the flag *and* the value, otherwise just checking if it'll overflow with `bool overflows = a > b` would do the job.

I like the idea, and the savings look promising. But either we wait until 6.x or release a separate version of `try*` called `unchecked*` (temptatively) so that people can "safely" use unchecked operations.

**Comment by Amxx** [comment]
> Your code snippet above is actually a good example of how wrong it is to rely on 0 alone as an indication of overflow. Anyone using such pattern **at this moment** (i.e., under the current implementation) is subjected to erroneous behavior.

I'm not sure it is a wrong example. 

If you only have to do some operation when the result is not 0 (lets say its a transfer and you only want to do it if value is > 0), then you can check the value, and if its not 0 you are sure its valid. If the value is 0, then you have to check if you are indeed with a no-op (in which case you do nothing) of if there was an overflow, and you need to revert.

Doing 
```
(uint256 result, bool success) = tryAdd(x, y);
if (result > 0) {
    // do stuff
} else require(success, "overflow");
```
instead of
```
(uint256 result, bool success) = tryAdd(x, y);
require(success, "overflow");
if (result > 0) {
    // do stuff
}
```
saves gas when there is not overflow and the result is not 0. If you expect that to be the majority of the cases, then it makes sens to use the first syntax. I would not condemn a dev that understands the details of `tryAdd` and use that to save its users some gas.

**Comment by barakman** [comment]
> > Your code snippet above is actually a good example of how wrong it is to rely on 0 alone as an indication of overflow. Anyone using such pattern **at this moment** (i.e., under the current implementation) is subjected to erroneous behavior.
> 
> I'm not sure it is a wrong example.
> 
> If you only have to do some operation when the result is not 0 (lets say its a transfer and you only want to do it if value is > 0), then you can check the value, and if its not 0 you are sure its valid. If the value is 0, then you have to check if you are indeed with a no-op (in which case you do nothing) of if there was an overflow, and you need to revert.
> 
> Doing
> 
> ```solidity
> (uint256 result, bool success) = tryAdd(x, y);
> if (result > 0) {
>     // do stuff
> } else require(success, "overflow");
> ```
> 
> instead of
> 
> ```solidity
> (uint256 result, bool success) = tryAdd(x, y);
> require(success, "overflow");
> if (result > 0) {
>     // do stuff
> }
> ```
> 
> saves gas when there is not overflow and the result is not 0. If you expect that to be the majority of the cases, then it makes sens to use the first syntax. I would not condemn a dev that understands the details of `tryAdd` and use that to save its users some gas.

Yes, I apologize, I mistaken your original code for this:
```
(uint256 result, bool success) = tryAdd(x, y);
if (result > 0) {
    // do stuff
} else revert("overflow");
```
Which is obviously erroneous handling for the case of `x` set to zero and `y` set to zero.
I've mistaken your `require` statement with a `revert` statement.
More precisely, I read it to hastily, eager to prove ,my point :)

But in either case, this just brings a 4th consideration (on top of the 3 which I have already mentioned), as to why it might be a win-win to move away from "embedded ambiguity" (sort to speak) in setting the result to zero when an overflow occurs.

Lastly, in order to avoid the risk of breaking this expected behavior for existing consumers, see my suggestion above.
TLDR: implement it in a new function on v5, and deprecate the current function on v6 (alongside appropriate documentation, including the differences between the two functions, in the contract AND in the release notes).
